x-pack/plugins/security/server/authentication/providers/anonymous.ts

```diff
@@ -0,0 +1,217 @@
+    );
+
+    // Having a `null` state means that provider was specifically called to do a logout, but when
+    // session isn't defined then provider is just being probed whether or not it can perform logout.
+    if (state === undefined) {
+      return DeauthenticationResult.notHandled();
+    }
+
+    return DeauthenticationResult.redirectTo(this.options.urls.loggedOut);
+  }
+
+  /**
+   * Returns HTTP authentication scheme (`Basic` or `ApiKey`) that's used within `Authorization`
+   * HTTP header that provider attaches to all successfully authenticated requests to Elasticsearch.
+   */
+  public getHTTPAuthenticationScheme() {
+    return isAPIKeyCredentials(this.credentials) ? 'apikey' : 'basic';
+  }
+
+  /**
+   * Tries to authenticate user request via configured username and password.
+   * @param request Request instance.
+   * @param credentials Credentials based on the username and password.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaUsernameAndPassword(
+    request: KibanaRequest,
+    { username, password }: UsernameAndPasswordCredentials,
+    state?: unknown
+  ) {
+    this.logger.debug('Trying to authenticate request via anonymous username and password.');
+
+    try {
+      const authHeaders = {
+        authorization: new HTTPAuthorizationHeader(
+          'Basic',
+          new BasicHTTPAuthorizationHeaderCredentials(username, password).toString()
+        ).toString(),
+      };
+
+      const user = await this.getUser(request, authHeaders);
+      this.logger.debug(
+        `Request to ${request.url.pathname}${request.url.search} has been authenticated.`
+      );
+      return AuthenticationResult.succeeded(user, {
+        authHeaders,
+        // Create session only if it doesn't exist yet, otherwise keep it unchanged.
+        state: state ? undefined : {},
+      });
+    } catch (err) {
+      this.logger.debug(
+        `Failed to authenticate request via anonymous username and password: ${err.message}`
+      );
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to authenticate user in via configured api key.
+   * @param request Request instance.
+   * @param credentials Credentials based on the API key.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaAPIKey(
+    request: KibanaRequest,
+    { apiKey, username }: APIKeyCredentials,
+    state?: unknown
+  ) {
+    this.logger.debug('Trying to authenticate request via API key.');
+
+    const authHeaders = {
+      authorization: new HTTPAuthorizationHeader('ApiKey', apiKey).toString(),
+    };
+    try {
+      const user = await this.getUser(request, authHeaders);
+      this.logger.debug(
+        `Request to ${request.url.pathname}${request.url.search} has been authenticated.`
+      );
+      return AuthenticationResult.succeeded(username ? { ...user, username } : user, {
```

Comment:
> question what is the use case for overriding the username of the user who created the API Key?

Yeah, I'm not entirely sure we need this either. But since initial UX proposal I've got implied that Profile link/page should stay even for anonymous users I wanted to display something there. And displaying the name of the user that created API key would also be confusing, so it's purely UI thing.

I'm wondering if we ES API keys API can be changed to allow us to specify username that should be associated with the API key.

>  They might also not understand the importance of the username -- for example, if you changed the displayed username to elastic, then (I think) you'll see all reports generated by the real elastic user. Likewise, the real elastic user will see all reports generated by their anonymous users.

Hmmm, I'd be really surprised if we base our checks on the username (ignoring authentication realms) :see_no_evil:  I'll double check if it's the case if we happen to keep this `username`.

> For example, I'm seeing audit logs for the canonical username as opposed to the overridden username. Was that intentional?

You mean Kibana audit logs? If so, then it's not intentional, and I'd say it's even concerning since audit logging should get username __only__ from `AuthenticatedUser`. That's a good catch, let me see where this coming from.

Reply:
> You mean Kibana audit logs? If so, then it's not intentional, and I'd say it's even concerning since audit logging should get username only from AuthenticatedUser. That's a good catch, let me see where this coming from.

Hmm, yeah, it's coming from the response of `shield.hasPrivileges` and we use it in the audit events. This doesn't look like something that'd be easy to change. Do you think we can just drop this idea with overriding username and see what our users say and then get back to it only if it becomes a problem?

Reply:
> Do you think we can just drop this idea with overriding username and see what our users say and then get back to it only if it becomes a problem?

++ yeah I think we should drop the username override for now, and re-evaluate if needed.

Reply:
Done.